### PR TITLE
[SSWG-0020] Distributed actors (cluster)

### PR DESCRIPTION
**Pull request by ktoso**
This proposes the inclusion of http://github.com/apple/swift-distributed-actors into the SSWG endorsed project list.

We propose it as **Sandbox** as we have not reached complete API stability yet, but would like to have more exposure that the SSWG listing can provide.

### CODE REVIEW

**Comment by ktoso** [comment]
I ran out of night today and will want to release a beta-3 before posting this on the forums tomorrow 👍

File: proposals/0020-distributed-actor-cluster.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+| **Proposed Maturity Level** | [Sandbox](https://github.com/swift-server/sswg/blob/main/process/incubation.md#process-diagram) |
+| **License** | [Apache 2.0](https://github.com/swift-server/swift-backtrace/blob/master/LICENSE.txt) |
+| **Dependencies** | [swift-nio](https://github.com/apple/swift-nio), [swift-log](https://github.com/apple/swift-log), [swift-metrics](https://github.com/apple/swift-metrics), [swift-service-discovery](https://github.com/apple/swift-service-discovery), [swift-cluster-membership](https://github.com/apple/swift-cluster-membership) |
+
+
+## Introduction
+
+Distributed actors are an extension of the "local only" actor model offered by Swift with its `actor` keyword.
+
+Distributed actors used by `distributed actor` keywords (and importing the `Distributed` module, provided by Swift), and enable the declaring of `distributed func` methods inside such actor. Such methods may then be invoked remotely, from other peers in a distributed actor system.
```
Comment (0xTim) [inline proposals/0020-distributed-actor-cluster.md]: Wording seems slightly off
```suggestion
Distributed actors used the `distributed actor` keywords (after importing the `Distributed` module, provided by Swift), and enable the declaring of `distributed func` methods inside such actor. Such methods may then be invoked remotely, from other peers in a distributed actor system.
```

File: proposals/0020-distributed-actor-cluster.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+The distributed actor *language feature* does not include any specific *runtime*, and only defines the language and semantic rules surrounding distributed actors. This library provides a feature-rich clustering server-side focused implementation of such runtime (i.e. a `DistributedActorSystem` implementation) for distributed actors.
+
+Clusters are a fundamental building block of server-side distributed systems, such as lobby systems for game backends, IoT service backends, orchestration systems and general control plane systems of distributed databases or other distributed systems.
+
+## Motivation
+
+Distributed actors are introduced as nominal types in Swift 5.7. Similarly to actors, they can be declared using the `distributed actor` pair of keywords. By themselves, they cannot really do anything – all the distributed actions such actor performs are actually handled an `ActorSystem` associated with given actor type.
```
Comment (0xTim) [inline proposals/0020-distributed-actor-cluster.md]:
```suggestion
Distributed actors are introduced as nominal types in Swift 5.7. Similar to actors, they can be declared using the `distributed actor` pair of keywords. By themselves, they cannot really do anything – all the distributed actions such actor performs are actually handled by an `ActorSystem` associated with given actor type. 
```

File: proposals/0020-distributed-actor-cluster.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+```
+
+Clusters are formed by “joining” other nodes. So the above created node has bound on port 7337, and if we had another node running on the same host but on port 8228 we could join in programmatically like this:
+
+```swift
+let other = Cluster.Endpoint(host: "127.0.0.1", port: 8228)
+system.cluster.join(endpoint: other)
+```
+
+The `join(endpoint:)` method returns immediately, however it is possible to `try await joined(endpoint:within:)` to suspend until the other node has been joined (or we failed doing so within the timeout).
```
Comment (0xTim) [inline proposals/0020-distributed-actor-cluster.md]:
```suggestion
The `join(endpoint:)` method returns immediately, however it is possible to use `try await joined(endpoint:within:)` to suspend until the other node has been joined (or we failed doing so within the timeout).
```

File: proposals/0020-distributed-actor-cluster.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+Once a cluster has been formed, distributed actors use the [Receptionist](https://apple.github.io/swift-distributed-actors/1.0.0-beta.2/documentation/distributedactors/receptionist) to discover check-in and discover distributed actors from other nodes. For example, we can write a distributed worker actor like this:
+
+```swift
+distributed actor Worker {
+    typealias ActorSystem = ClusterSystem
+    
+    init(actorSystem: ActorSystem) async {
+        self.actorSystem = actorSystem
+        await system.receptionist.checkIn(worker, with: .workers)
```
Comment (0xTim) [inline proposals/0020-distributed-actor-cluster.md]: where is `system` from here - is it a singleton?

File: proposals/0020-distributed-actor-cluster.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+extension DistributedReception.Key {
+    static var workers: DistributedReception.Key<Worker> {
+        "workers"
+    }
+}
+```
+
+And have all workers that are initialized on any node that has joined the cluster by using the [receptionist listing](https://apple.github.io/swift-distributed-actors/1.0.0-beta.2/documentation/distributedactors/receptionist#Receptionist-Listings):
```
Comment (0xTim) [inline proposals/0020-distributed-actor-cluster.md]: This sentence needs refining

**Review by 0xTim — COMMENTED** [review]
Some minor grammar suggestions but overall looks good!

File: proposals/0020-distributed-actor-cluster.md (hunk 4 above)
Comment (ktoso) [reply proposals/0020-distributed-actor-cluster.md]: typo, thanks!

It is supposed to be `actorSystem`; my muscle memory won here hehe... I often call it system in the lib

```suggestion
        await actorSystem.receptionist.checkIn(worker, with: .workers) 
```

File: proposals/0020-distributed-actor-cluster.md (hunk 5 above)
Comment (ktoso) [reply proposals/0020-distributed-actor-cluster.md]: Thanks will do

```suggestion
Such distributed actors are then created using normal initialization on nodes within the cluster. Each such node can then use the [receptionist listing](https://apple.github.io/swift-distributed-actors/1.0.0-beta.3/documentation/distributedactors/receptionist#Receptionist-Listings) API to obtain all workers from all other nodes, like this:
```

**Review by adam-fowler — APPROVED** [review]
Looks good

**Review by kmahar — APPROVED** [review]

**Comment by kmahar** [comment]
I am going to merge this and kick off the proposal review thread on the forums

**Comment by ktoso** [comment]
Thank you, @kmahar !

**Pull request merged**